### ISSUE TITLE
player not destroy  on location changed

### ISSUE THREAD

**Issue by sajjad-ser** [jaruba/wcjs-player#4]
```html
    <div id="player"></div>
    <a href="index.htm">index.htm</a>
    <script>
        var wjs = require("wcjs-player");
        var p = new wjs("#player");
        var player = p.addPlayer();
        player.addPlaylist("http://localhost:8000/a.avi");
        player.play();
    </script>
```
when i click on anchor the player still work on background

**Comment by RSATom** [comment]
It could be JavaScript Garbage Collector issue. It has some delay before start killing not used anymore objects. It's better call `player.stop()` if you don't like this behavior.

**Comment by RSATom** [comment]
It's possible `delete player;` at some point could help with this issue, but I have some doubts...

**Comment by jaruba** [comment]
Try this:

    <a href="index.htm" onClick="player.stop()">index.htm</a>

**Comment by sajjad-ser** [comment]
I use ```player.stop()``` and ```delete player``` it's work but when i come back to the ```player.htm``` player can not initialize and console error is
````
Uncaught TypeError: Cannot read property 'firstChild' of nullD:\nw\p\node_modules\wcjs-player\index.js:43 wjs2.html:32 (anonymous function)
````

**Comment by sajjad-ser** [comment]
player.htm
```html
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title></title>
    <style type="text/css" >
    html,body{
        height: 100%;
        width: 100%;
    }
    #player{
        height: 90%;
        width: 100%;
    }
    </style>
</head>
<body>
    <div id="player"></div>
    <a href="index.htm">2222</a>
    <script>
        var wjs = require("wcjs-player");
        var player = new wjs("#player").addPlayer();
        player.addPlaylist("http://localhost:8000/a.avi");
        player.play();
    </script>
</body>
</html>
```

index.htm
```html
<!DOCTYPE html>
<html>

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title></title>
</head>

<body>
<a href="player.htm"> player.htm</a>

</body>

</html>

```

**Comment by jaruba** [comment]
Thanks, I'll try to reproduce the issue.

**Comment by jaruba** [comment]
I guess you'll have to clear the cache before using ``require()`` on the new page.

**player.html**

	// clear NW.js require cache
	for(module in global.require.cache) {
		if(global.require.cache.hasOwnProperty(module)) {
			delete global.require.cache[module];
		}
	}

	var wjs = require("wcjs-player");
	var player = new wjs("#player").addPlayer();
	player.addPlaylist("http://localhost:8000/a.avi");
	player.play();

**Comment by RSATom** [comment]
@jaruba, do you do any initialization stuff on `require("wcjs-player");` call?

**Comment by jaruba** [comment]
I have a few global variable that are declared and I inject the css stylesheet in the page. But there are many more issues with this..

For example I use jQuery, and without removing the cache from NW.js programatically, jQuery will still only be able to access the elements of the previous page (where ``require`` was first called), which don't even exist anymore.. It's necessary to empty the cache on every new page load.

This is not really a design flaw.. imho NW.js / Electron apps should never use more then one page, they should dynamically create all the elements that are needed to be seen in the window..

**Comment by RSATom** [comment]
Is there any other modules which have the same problem?

**Comment by jaruba** [comment]
I never used multi-page apps before, so I can't answer that question..

**Comment by RSATom** [comment]
I mean, is there any npm modules which has problem with multiple pages and require clear cache?

**Comment by jaruba** [comment]
I understood the question and answered it above as best I could. :)

**Comment by RSATom** [comment]
Ok, thanks :)

**Comment by RSATom** [comment]
so, we could consider jQuery npm package has a bug?

**Comment by jaruba** [comment]
This is not a bug.. I don't think there's anything I can do to "fix" this inside the module. It's the normal node.js behavior..

**Comment by jaruba** [comment]
Node.js just naturally presumes that the modules don't need to interact with the HTML page, if I would use a torrent downloading module I would then see this feature as useful, because I would want that process to survive between page changes..

But I think all the modules that need to interact with HTML page are subjective to this problem, because they remain linked to the previous HTML page where the module was first called.

**Comment by RSATom** [comment]
so, you think resetting npm cache is right solution? - Ok, will not argue, since I don't know node.js internals well enough.

**Comment by jaruba** [comment]
It could be tweaked a bit to only remove the ``wcjs-player`` module and all of it's dependencies:

    // clear wcjs-player from require cache
    for(module in global.require.cache) {
        if(global.require.cache.hasOwnProperty(module) && module.indexOf("wcjs-player") > -1) {
            delete global.require.cache[module];
        }
    }

**Comment by jaruba** [comment]
Fixed in ``v0.3.6`` with: https://github.com/jaruba/wcjs-player/commit/15b2a61405a3fb9da2e9ce71b83e38470f14b9db

**Comment by jaruba** [comment]
With this change: https://github.com/jaruba/wcjs-player/commit/19b48039c750f8f16f369aa0355bd09e73825ed1 it should now also stop all players when the page is changed.
